_posts/2019-04-23-js-recon-is-no-more.md

```diff
@@ -0,0 +1,147 @@
+---
+layout: post
+title: "JS-Recon is no more!"
+date: "2019-04-23 17:47:21 +0530"
+tag:
+  - security
+  - attack
+  - vulnerability
+  - javascript
+  - jsrecon
+---
+
+In 2010, Andlabs discovered an attack which fingerprints open TCP ports in
+client workstations. You can read [this blog post][andlabs_blogpost] which
+describes the details of this method. JS-Recon was a tool implementing this
+attack on its clients to prove the danger of this vulnerability. Unfortunately,
+JS-Recon is no longer available on the aforementioned link for unknown reasons.
+I tried my best to find the source code of this tool, but I finished with no
+results.  Because the source code of the tool is not available, the only way to
+confirm the possibility of this attack was to reconstruct it from the steps
+mentioned in that blog post.
+
+In this post I will share my experience of rebuilding this attack. Because this
+attack is from the front-end side, knowledge of basic Javascript API is expected
+from the reader. The original blog post does not include any code samples. For
+the benefit of the reader, I have prepared small code snippets and attached them
+with related sections. I expect you run code samples in the developer console of
+your browser.
+
+### Glossary
+
+* **Empty / Available port**: A port where no service is running.
+* **Non-empty / Occupied port**: A port where some service is running.
+* **XHR**: A short form of [XML Http Request][mdn_xhr].
+* **Socket**: A TCP/IP raw socket.
+
+According to that blog post, If I write a Javascript code to open an XHR to
```

Comment:
The XHR API distinguishes between opening (initializing) and sending (triggering/delivering) a request, so be more precise here.